### ISSUE TITLE
"Multisampled" textures with sampleCount=1?

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1226]
A long, long time ago we decided that all textures would be conceptually "multisample" and some usages would simply require `sampleCount=1`. This seems to be what Vulkan has done. However, Metal doesn't make this decision and has `MTLTextureType2D` and `MTLTextureType2DMultisample`.

This becomes a problem with argument buffers because they also distinguish the two types. I haven't looked at argument buffers closely but @kvark says you can't just take a 2d texture and put it in a 2d-multisample argument buffer slot.

In particular, in bind group layouts, this means you can't bind a texture with sampleCount=1 to a `"multisampled-texture"` binding point. This really seems like something that should be possible. But with the constraints of Metal (and possibly D3D12?), it seems necessary for texture creation to distinguish non-multisample textures from single-sampled-multisample textures.

**Comment by kvark** [comment]
To clarify, there may be use cases where a texture may have some number of samples (or 1), and the shader code just loads from it anyway. In this case, it could make sense for us to support this case with the same binding type.
The restriction on Metal (with argument buffers) needs to be double-checked. Vulkan and D3D12 seem fine.

**Comment by Kangz** [comment]
I don't think we can talk about this in the group before we have investigated the behavior in Metal. Putting back into "Needs investigation".

**Comment by jdashg** [comment]
Maybe we want `u32? sampleCount`?

**Comment by kainino0x** [comment]
Would remove the defaulting, to make it optional, rather than making it nullable, but yeah

**Comment by kainino0x** [comment]
This seems like a pretty high priority issue as it would be a fundamental change. I still know very little about argument buffers and it would add some overhead for me to investigate it. Anyone up to investigate it?

cc @austinEng @litherum just in case

**Comment by kvark** [comment]
Hold on, this isn't meant to be "fundamental change" at all. We are talking about allowing sample count = 1 for multi-sampled textures. We aren't talking about treating non-multisampled textures equivalent to MSAA count = 1.
So we can easily postpone this till later after MVP.

FWIW, I did a quick test on Metal, and it doesn't like sample count = 1:
> -[MTLTextureDescriptorInternal validateWithDevice:]:1248: failed assertion `Texture Descriptor Validation
MTLTextureDescriptor sampleCount must be > 1 for multisample textures.
MTLTextureDescriptor: Multisample textures cannot be allocated with MTLStorageModeShared or MTLStorageModeManaged on this device.

**Comment by Kangz** [comment]
@kvark I guess the other question is whether you can encode a non-multisampled texture to a multisample texture slot in a Metal argument buffer.

Maybe it's fine to say that we'll never have 1-sample multisampled textures. In particular that's a constraint of Vulkan too:

> If the image was created with VkImageCreateInfo::samples equal to VK_SAMPLE_COUNT_1_BIT, the instruction must have MS = 0.

Meaning that if you use a `texture_2d_ms` it must have been created with samples > 1.

**Comment by kvark** [comment]
@kainino0x thank you for clarification! I didn't understand the concern fully, now that I think of it. So you were worried that if we ever want to support count=1 MSAA textures, we'd not have an API to express this.
Given the limitation of Metal and Vulkan here, seems pretty clear that we shouldn't try to support this in the future.
